src/vm/eventpipesession.cpp

```diff
@@ -44,23 +84,151 @@ EventPipeSession::~EventPipeSession()
     }
     CONTRACTL_END;
 
-    if(m_pProviderList != NULL)
+    // TODO: Stop streaming thread? Review synchronization.
+    // CrstHolder _crst(&m_lock);
+
+    delete m_pProviderList;
+
+    // FIXME: Verify DeAllocateBuffers is called when destructing the buffer manager.
+    m_pBufferManager->DeAllocateBuffers();
+    delete m_pBufferManager;
+    delete m_pFile;
+}
+
+bool EventPipeSession::HasIpcStreamingStarted()
+{
+    // CrstHolder _crst(&m_lock);
+    return m_pIpcStreamingThread != nullptr ? m_pIpcStreamingThread->HasStarted() : false;
+}
+
+void EventPipeSession::DestroyIpcStreamingThread()
+{
+    // CrstHolder _crst(&m_lock);
+
+    if (!m_ipcStreamingEnabled)
+        return;
+
+    if (m_pIpcStreamingThread != nullptr)
+        ::DestroyThread(m_pIpcStreamingThread);
+
+    m_pIpcStreamingThread = nullptr;
+
+    // Signal Disable() that the thread has been destroyed.
+    m_threadShutdownEvent.Set();
+}
+
+static DWORD WINAPI ThreadProc(void *args)
+{
+    CONTRACTL
     {
-        delete m_pProviderList;
-        m_pProviderList = NULL;
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(args != nullptr);
     }
+    CONTRACTL_END;
+
+    // TODO: Review Andrew's update/PR on SamplerProfiler
+
+    if (args == nullptr)
+        return 1;
+
+    EventPipeSession *const pEventPipeSession = reinterpret_cast<EventPipeSession *>(args);
+    if (pEventPipeSession->GetSessionType() != EventPipeSessionType::IpcStream)
+        return 1;
+
+    EX_TRY
+    {
+        if (pEventPipeSession->HasIpcStreamingStarted())
+        {
+            bool fSuccess = true;
+            while (pEventPipeSession->IsIpcStreamingEnabled() && fSuccess)
+            {
+                {
+                    // Switch to pre-emptive mode so that this thread doesn't starve the GC
+                    GCX_PREEMP();
```

Comment:
Is this just copy-and-paste from SampleProfiler?

I am not sure why are we doing it - the contract indicates this thread is MODE_PREEMPTIVE, not sure why we need to do an explicit switch here.